mmpose/models/heads/heatmap_heads/cid_head.py

```diff
@@ -0,0 +1,727 @@
+                data samples
+            test_cfg (dict): The runtime config for testing process. Defaults
+                to {}
+
+        Returns:
+            Union[InstanceList | Tuple[InstanceList | PixelDataList]]: If
+            ``test_cfg['output_heatmap']==True``, return both pose and heatmap
+            prediction; otherwise only return the pose prediction.
+
+            The pose prediction is a list of ``InstanceData``, each contains
+            the following fields:
+
+                - keypoints (np.ndarray): predicted keypoint coordinates in
+                    shape (num_instances, K, D) where K is the keypoint number
+                    and D is the keypoint dimension
+                - keypoint_scores (np.ndarray): predicted keypoint scores in
+                    shape (num_instances, K)
+
+            The heatmap prediction is a list of ``PixelData``, each contains
+            the following fields:
+
+                - heatmaps (Tensor): The predicted heatmaps in shape (K, h, w)
+        """
+        if test_cfg.get('flip_test', False):
+            assert isinstance(feats, list) and len(feats) == 2
+
+            feats_flipped = flip_heatmaps(
+                self._transform_inputs(feats[1]), shift_heatmap=False)
+            feats = torch.cat(
+                (self._transform_inputs(feats[0]), feats_flipped))
+        else:
+            feats = self._transform_inputs(feats)
+
+        instance_info = self.iia_module.forward_test(feats, test_cfg)
+        instance_feats, instance_coords, instance_scores = instance_info
+        if len(instance_coords) > 0:
+            instance_indices = torch.zeros(
+                instance_coords.size(0), dtype=torch.long, device=feats.device)
+            if test_cfg.get('flip_test', False):
+                instance_coords = torch.cat((instance_coords, instance_coords))
+                instance_indices = torch.cat(
+                    (instance_indices, instance_indices + 1))
+            instance_heatmaps = self.gfd_module(feats, instance_feats,
+                                                instance_coords,
+                                                instance_indices)
+            if test_cfg.get('flip_test', False):
+                flip_indices = batch_data_samples[0].metainfo['flip_indices']
+                instance_heatmaps, instance_heatmaps_flip = torch.chunk(
+                    instance_heatmaps, 2, dim=0)
+                instance_heatmaps_flip = \
+                    instance_heatmaps_flip[:, flip_indices, :, :]
+                instance_heatmaps = (instance_heatmaps +
+                                     instance_heatmaps_flip) / 2.0
+            instance_heatmaps = smooth_heatmaps(
+                instance_heatmaps, test_cfg.get('blur_kernel_size', 3))
+
+            preds = self.decode((instance_heatmaps, instance_scores[:, None]))
+            preds = InstanceData.cat(preds)
+            preds.keypoints[..., :2] += 1.5
```

Comment:
why 1.5?

Reply:
For example, consider x. Its range in the input image space is [0, w-1], but in the heatmap space, it is [0, w-4] after being rescaled. The shift is (4-1)/2=1.5. 

However, I have found that using `preds.keypoints[..., :2] += 2` results in higher accuracy. I will fix this in a later commit. Do you have any idea about the reason?

Reply:
0.5 * 4 (down-sampling scale) = 2?